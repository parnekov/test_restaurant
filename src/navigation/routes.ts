/* MODULES */
import {NavigatorScreenParams} from '@react-navigation/native';
import {createNativeStackNavigator} from '@react-navigation/native-stack';

export enum SCREEN_NAMES {
  RESTAURANTS = 'Restaurants',
  TABS_ORDERS = 'Tabs',
  ALL_ORDERS = 'All orders',
  MY_ORDER = 'My order',
}

export type StackParamList = {
  [SCREEN_NAMES.RESTAURANTS]: undefined;
  [SCREEN_NAMES.TABS_ORDERS]: undefined;
};

export type TabsParamList = {
  [SCREEN_NAMES.ALL_ORDERS]: undefined;
  [SCREEN_NAMES.MY_ORDER]: undefined;
};

export enum ROOT_ROUTES {
  RESTAURANT_STACK = 'RESTAURANT_STACK',
  TABS_STACK = 'TABS_STACK',
}

export type RootStackParamList = {
  [SCREEN_NAMES.RESTAURANTS]: NavigatorScreenParams<StackParamList>;
  [SCREEN_NAMES.TABS_ORDERS]: NavigatorScreenParams<TabsParamList>;
};

export const Stack = createNativeStackNavigator<StackParamList>();
export const Tabs = createNativeStackNavigator<TabsParamList>();
