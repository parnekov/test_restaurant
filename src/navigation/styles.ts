/* REACT */
import {StyleSheet} from 'react-native';
/* UTILS */
import {COLORS, FONT_SIZES} from '../utils/theme';

export default StyleSheet.create({
  tabBarLabelStyle: {
    textTransform: 'none',
    fontSize: FONT_SIZES.BIG,
    fontWeight: '600',
    color: COLORS.TEXT_SUBTITLE,
  },
  tabBarIndicatorStyle: {
    backgroundColor: COLORS.PRIMARY_LIGHT,
  },
});
