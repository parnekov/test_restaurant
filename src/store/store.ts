/* MODULES */
import {configureStore} from '@reduxjs/toolkit';
/* REDUCERS */
import OrdersReducer from './slices/ordersSlice';

export const store = configureStore({
  reducer: {
    order: OrdersReducer,
  }, // @Note: As there is an error with empty object use a Reducer as a stub.
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: false, // fix of "A non-serializable value was detected in the state" error
    }),
});

// @Note: infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
