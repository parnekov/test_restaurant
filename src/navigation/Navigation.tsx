/* REACT */
import React from 'react';
/* COMPONENTS */
import {NavigationContainer} from '@react-navigation/native';
import StackNavigator from './StackNavigator';

const Navigation = () => (
  <NavigationContainer>
    <StackNavigator />
  </NavigationContainer>
);

export default Navigation;
