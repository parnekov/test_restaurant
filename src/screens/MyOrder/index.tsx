/* REACT */
import React, {useCallback, useEffect, useLayoutEffect, useMemo} from 'react';
import {View, Text, SectionList} from 'react-native';
/* MODULES */
import {useNavigation} from '@react-navigation/native';
/* STORE */
import {useDispatch} from 'react-redux';
import {useAppSelector} from '../../store/hooks';
import {selectMyOrder} from '../../store/slices/ordersSlice';
import {
  checkUncheckMyOrderItems,
  getOrdersThunk,
  changeAmountMyOrderItem,
  selectRestaurantId,
  selectRestaurantQrCodeName,
  postOrdersThunk,
} from '../../store/slices/ordersSlice';
/* COMPONENTS */
import OrderListItem from '../../components/OrderListItem';
import CheckBox from '../../components/CheckBox';
import OrderScreenWrapper from '../../components/OrderScreenWrapper';
import Total from '../../components/Total';
import PayButton from '../../components/PayButton';
/* UTILS */
import {COLORS} from '../../utils/theme';
import {makeMyOrderBodyToPost} from '../../utils/builders';
import {ORDER_STATUS} from '../../utils/constants';
import {Order, OrderItem} from '../../utils/types';
import {SCREEN_NAMES} from '../../navigation/routes';
/* STYLES */
import styles from './styles';

const testID = SCREEN_NAMES.MY_ORDER;

const MyOrder = () => {
  const dispatch = useDispatch();
  const restaurantId = useAppSelector(selectRestaurantId);
  const qrCodeName = useAppSelector(selectRestaurantQrCodeName);

  const navigation = useNavigation();
  const myOrder = useAppSelector(selectMyOrder);
  const isSelectedAll = myOrder?.orderItems.every(
    item => item.checked,
  ) as boolean;

  useLayoutEffect(() => {
    navigation.setOptions({
      tabBarLabel: SCREEN_NAMES.MY_ORDER,
    });
  }, [navigation]);

  useEffect(() => {
    // @ts-ignore
    dispatch(getOrdersThunk());
  }, [dispatch]);

  const checkUncheckAll = useCallback(
    (isCheck: boolean) =>
      dispatch(checkUncheckMyOrderItems({checked: isCheck})),
    [dispatch],
  );

  const makeData = useMemo(() => {
    const newItems = myOrder?.orderItems.filter(
      item => item.status === ORDER_STATUS.New,
    );

    const inProgressItems = myOrder?.orderItems.filter(
      item => item.status === ORDER_STATUS.InProgress,
    );

    const readyItems = myOrder?.orderItems.filter(
      item => item.status === ORDER_STATUS.Ready,
    );

    const data = [];

    if (newItems?.length) {
      data.push({
        title: 'New',
        data: newItems,
      });
    }

    if (inProgressItems?.length) {
      data.push({
        title: 'In Progress',
        data: inProgressItems,
      });
    }

    if (readyItems?.length) {
      data.push({
        title: 'Ready',
        data: readyItems,
      });
    }

    return data;
  }, [myOrder]);

  const totalPrice = useCallback(() => {
    return (
      myOrder?.orderItems.reduce((acc, curr) => {
        return curr.checked ? acc + curr.amount * curr.price : acc;
      }, 0) || 0
    );
  }, [myOrder]);

  const onChangeCheckOrderItem = useCallback(
    ({check, itemId}: {check: boolean; itemId: string}) => {
      dispatch(checkUncheckMyOrderItems({checked: check, itemId}));
    },
    [dispatch],
  );

  const onChangeAmountItem = useCallback(
    ({updatedAmount, itemId}: {updatedAmount: number; itemId: string}) => {
      dispatch(changeAmountMyOrderItem({updatedAmount, itemId}));
    },
    [dispatch],
  );

  const renderItem = useCallback(
    ({item}: {item: OrderItem; index: number}) => (
      <OrderListItem
        orderItem={item}
        onChangeCheckbox={onChangeCheckOrderItem}
        onChangeAmount={onChangeAmountItem}
      />
    ),
    [onChangeCheckOrderItem, onChangeAmountItem],
  );

  const renderSectionHeader = useCallback(
    ({section: {title}}: {section: {title: string}}) => (
      <Text style={styles.sectionHeader}>{title}</Text>
    ),
    [],
  );

  const amountItems = useMemo(() => {
    return myOrder?.orderItems?.reduce((acc, curr) => {
      return curr.checked && !curr?.isPaid ? acc + curr.amount : acc;
    }, 0);
  }, [myOrder]);

  const renderSectionFooter = useCallback(
    () => <Total amountItems={amountItems as number} price={totalPrice()} />,
    [amountItems, totalPrice],
  );

  const payOrders = () => {
    const body = makeMyOrderBodyToPost(myOrder as Order);
    // @ts-ignore
    dispatch(postOrdersThunk({restaurantId, qrCodeName, body}));
  };

  return (
    <OrderScreenWrapper testId={`${testID}-container`}>
      <SectionList
        testID={`${testID}-list`}
        contentContainerStyle={styles.sectionContainer}
        ListHeaderComponent={
          <CheckBox
            checked={isSelectedAll}
            title={isSelectedAll ? 'Unselect all' : 'Select all'}
            onChange={checkUncheckAll}
            style={styles.checkbox}
            checkedColor={COLORS.GRADIENT_TOP}
            testID={`${testID}-list-header-checkbox`}
          />
        }
        sections={makeData}
        keyExtractor={(item, index) => item.id + index}
        renderItem={renderItem}
        renderSectionHeader={renderSectionHeader}
        ListFooterComponent={renderSectionFooter}
      />
      <View style={styles.payButtonContainer}>
        <PayButton
          amountItems={amountItems as number}
          price={totalPrice() as number}
          onPress={payOrders}
          testID={`${testID}-pay-button`}
        />
      </View>
    </OrderScreenWrapper>
  );
};

export default MyOrder;
