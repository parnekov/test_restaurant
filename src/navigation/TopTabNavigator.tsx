/* REACT */
import React, {useLayoutEffect} from 'react';
/* MODULES */
import {createMaterialTopTabNavigator} from '@react-navigation/material-top-tabs';
import {useNavigation} from '@react-navigation/native';
/* COMPONENTS */
import AllOrders from '../screens/AllOrders';
import MyOrder from '../screens/MyOrder';
import HeaderTitle from '../components/HeaderTitle';
import HeaderRight from '../components/HeaderRight';
/* UTILS */
import {SCREEN_NAMES} from './routes';
/* STYLES */
import styles from './styles';

const Tab = createMaterialTopTabNavigator();

const TopTabNavigator = () => {
  const navigation = useNavigation();

  useLayoutEffect(() => {
    navigation.setOptions({
      headerTitle: HeaderTitle,
      headerRight: HeaderRight,
    });
  }, [navigation]);

  return (
    <Tab.Navigator
      screenOptions={{
        tabBarLabelStyle: styles.tabBarLabelStyle,
        tabBarIndicatorStyle: styles.tabBarIndicatorStyle,
      }}>
      <Tab.Screen
        name={SCREEN_NAMES.MY_ORDER}
        component={MyOrder}
        listeners={{
          tabPress: () => navigation.setOptions({headerTitle: HeaderTitle}),
        }}
      />
      <Tab.Screen
        name={SCREEN_NAMES.ALL_ORDERS}
        component={AllOrders}
        listeners={{
          tabPress: () => {
            navigation.setOptions({
              headerTitle: () => <HeaderTitle isNoOrder />,
            });
          },
        }}
      />
    </Tab.Navigator>
  );
};

export default TopTabNavigator;
