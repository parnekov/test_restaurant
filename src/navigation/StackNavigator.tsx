import React from 'react';
/* MODULES */
import {createNativeStackNavigator} from '@react-navigation/native-stack';
/* COMPONENTS */
import TopTabNavigator from './TopTabNavigator';
import Restaurants from '../screens/Restaurants';
/* UTILS */
import {SCREEN_NAMES} from './routes';
import {COLORS} from '../utils/theme';

const Stack = createNativeStackNavigator();

const StackNavigator = () => (
  <Stack.Navigator
    screenOptions={{
      headerBackTitleVisible: false,
      headerTintColor: COLORS.PRIMARY,
    }}>
    <Stack.Screen name={SCREEN_NAMES.RESTAURANTS} component={Restaurants} />
    <Stack.Screen name={SCREEN_NAMES.TABS_ORDERS} component={TopTabNavigator} />
  </Stack.Navigator>
);

export default StackNavigator;
