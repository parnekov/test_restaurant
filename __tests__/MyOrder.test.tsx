/* REACT */
import React from 'react';
/* MODULES */
import {ReactTestInstance} from 'react-test-renderer';
import {render, fireEvent, waitFor} from '@testing-library/react-native';
/* STORE */
import {Provider} from 'react-redux';
import {store} from '../src/store/store';
/* COMPONENTS */
import {NavigationContainer} from '@react-navigation/native';
import MyOrder from '../src/screens/MyOrder';
/* CONSTANTS */
import {SCREEN_NAMES} from '../src/navigation/routes';

const testID = SCREEN_NAMES.MY_ORDER;

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      setOptions: jest.fn(),
    }),
  };
});

const ComponentToRender = (
  <Provider store={store}>
    <NavigationContainer>
      <MyOrder />
    </NavigationContainer>
  </Provider>
);

describe('MyOrder screen tests', () => {
  it('MyOrder renders correctly', async () => {
    const {queryByTestId} = render(ComponentToRender);

    expect(queryByTestId(`${testID}-container`)).toBeTruthy();
    expect(queryByTestId(`${testID}-list`)).toBeTruthy();
    expect(queryByTestId(`${testID}-list-header-checkbox`)).toBeTruthy();
    expect(queryByTestId(`${testID}-pay-button`)).toBeTruthy();
  });

  it('MyOrder matches snapshot', () => {
    const component = render(ComponentToRender).toJSON();
    expect(component).toMatchSnapshot();
  });
});
