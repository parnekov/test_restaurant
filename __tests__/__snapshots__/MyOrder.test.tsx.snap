// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyOrder screen tests MyOrder matches snapshot 1`] = `
<BVLinearGradient
  colors={
    [
      4294955419,
      4294949277,
      4294947762,
    ]
  }
  endPoint={
    {
      "x": 0,
      "y": 1,
    }
  }
  locations={
    [
      0,
      0.4,
      0.7,
    ]
  }
  startPoint={
    {
      "x": 0,
      "y": 0.15,
    }
  }
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
  testID="My order-container"
>
  <RCTScrollView
    ListFooterComponent={[Function]}
    ListHeaderComponent={
      <CheckBox
        checkedColor="#FFD19B"
        onChange={[Function]}
        style={
          {
            "marginBottom": 16,
          }
        }
        testID="My order-list-header-checkbox"
        title="Select all"
      />
    }
    contentContainerStyle={
      {
        "padding": 16,
        "paddingBottom": 128,
      }
    }
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    testID="My order-list"
  >
    <View>
      <View
        collapsable={false}
        onLayout={[Function]}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "marginBottom": 16,
              "opacity": 1,
            }
          }
          testID="My order-list-header-checkbox"
        >
          <View
            style={
              [
                {
                  "borderColor": "#000000",
                  "borderRadius": 2,
                  "borderWidth": 1.5,
                  "height": 16,
                  "marginRight": 8,
                  "width": 16,
                },
                {
                  "backgroundColor": "#FFFFFF",
                },
              ]
            }
          >
            <RNSVGSvgView
              align="xMidYMid"
              bbHeight={16}
              bbWidth={16}
              color="#FFFFFF"
              focusable={false}
              height={16}
              meetOrSlice={0}
              minX={0}
              minY={0}
              size={16}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  {
                    "flex": 0,
                    "height": 16,
                    "width": 16,
                  },
                ]
              }
              tintColor="#FFFFFF"
              vbHeight={32}
              vbWidth={32}
              width={16}
            >
              <RNSVGGroup
                fill={
                  {
                    "payload": 4278190080,
                    "type": 0,
                  }
                }
              >
                <RNSVGPath
                  d="m5 16.577 2.194-2.195 5.486 5.484L24.804 7.743 27 9.937l-14.32 14.32z"
                  fill={
                    {
                      "payload": 4294967295,
                      "type": 0,
                    }
                  }
                  propList={
                    [
                      "fill",
                    ]
                  }
                />
              </RNSVGGroup>
            </RNSVGSvgView>
          </View>
          <Text
            style={
              {
                "color": "#60324E",
                "fontSize": 16,
                "fontWeight": "600",
              }
            }
          >
            Select all
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 2,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 16,
              "padding": 16,
            }
          }
        >
          <Text
            style={
              {
                "color": "#532C45",
                "fontSize": 18,
                "fontWeight": "700",
              }
            }
          >
            Total: 
          </Text>
          <Text
            style={
              {
                "color": "#532C45",
                "fontSize": 18,
                "fontWeight": "700",
              }
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#532C45",
                    "fontSize": 14,
                    "fontWeight": "700",
                  },
                ]
              }
            />
             
            <Text
              style={
                {
                  "color": "#532C45",
                  "fontSize": 14,
                  "fontWeight": "700",
                }
              }
            >
              / 
            </Text>
            <Text
              style={
                {
                  "color": "#532C45",
                  "fontSize": 14,
                  "fontWeight": "500",
                }
              }
            >
              CHF
            </Text>
             
            <Text>
              0.00
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "bottom": 46,
        "paddingHorizontal": 16,
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#B0B0B0",
          "borderRadius": 32,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "opacity": 1,
          "padding": 16,
        }
      }
      testID="My order-pay-button"
    >
      <Text
        style={
          {
            "color": "#FFFFFF",
            "fontSize": 18,
            "fontWeight": "700",
          }
        }
      >
        PAY 
      </Text>
      <Text
        style={
          {
            "color": "#FFFFFF",
            "fontSize": 18,
            "fontWeight": "700",
          }
        }
      >
        <Text />
         
        <Text>
          / 
        </Text>
        <Text>
          0.00
        </Text>
         
        <Text>
          CHF
        </Text>
      </Text>
    </View>
  </View>
</BVLinearGradient>
`;
